Lib/fontTools/ttLib/tables/_g_l_y_f.py

```diff
@@ -254,57 +260,41 @@ def __len__(self):
 		assert len(self.glyphOrder) == len(self.glyphs)
 		return len(self.glyphs)
 
-	def getPhantomPoints(self, glyphName, ttFont, defaultVerticalOrigin=None):
+	def getPhantomPoints(self, glyphName, hMetrics, vMetrics=None):
 		"""Compute the four "phantom points" for the given glyph from its bounding box
 		and the horizontal and vertical advance widths and sidebearings stored in the
 		ttFont's "hmtx" and "vmtx" tables.
 
-		If the ttFont doesn't contain a "vmtx" table, the hhea.ascent is used as the
-		vertical origin, and the head.unitsPerEm as the vertical advance.
+		'hMetrics' should be ttFont['hmtx'].metrics.
 
-		The "defaultVerticalOrigin" (Optional[int]) is needed when the ttFont contains
-		neither a "vmtx" nor an "hhea" table, as may happen with 'sparse' masters.
-		The value should be the hhea.ascent of the default master.
+		'vMetrics' should be ttFont['vmtx'].metrics if there is "vmtx" or None otherwise.
+		If there is no vMetrics passed in, vertical phantom points are set to the zero coordinate.
 
 		https://docs.microsoft.com/en-us/typography/opentype/spec/tt_instructing_glyphs#phantoms
 		"""
 		glyph = self[glyphName]
-		assert glyphName in ttFont["hmtx"].metrics, ttFont["hmtx"].metrics
-		horizontalAdvanceWidth, leftSideBearing = ttFont["hmtx"].metrics[glyphName]
 		if not hasattr(glyph, 'xMin'):
 			glyph.recalcBounds(self)
+
+		horizontalAdvanceWidth, leftSideBearing = hMetrics[glyphName]
 		leftSideX = glyph.xMin - leftSideBearing
 		rightSideX = leftSideX + horizontalAdvanceWidth
-		if "vmtx" in ttFont:
-			verticalAdvanceWidth, topSideBearing = ttFont["vmtx"].metrics[glyphName]
+
+		if vMetrics:
+			verticalAdvanceWidth, topSideBearing = vMetrics[glyphName]
 			topSideY = topSideBearing + glyph.yMax
+			bottomSideY = topSideY - verticalAdvanceWidth
 		else:
-			# without vmtx, use ascent as vertical origin and UPEM as vertical advance
-			# like HarfBuzz does
-			verticalAdvanceWidth = ttFont["head"].unitsPerEm
-			if "hhea" in ttFont:
-				topSideY = ttFont["hhea"].ascent
-			else:
-				# sparse masters may not contain an hhea table; use the ascent
-				# of the default master as the vertical origin
-				if defaultVerticalOrigin is not None:
-					topSideY = defaultVerticalOrigin
-				else:
-					log.warning(
-						"font is missing both 'vmtx' and 'hhea' tables, "
-						"and no 'defaultVerticalOrigin' was provided; "
-						"the vertical phantom points may be incorrect."
-					)
-					topSideY = verticalAdvanceWidth
-		bottomSideY = topSideY - verticalAdvanceWidth
+			bottomSideY = topSideY = 0
+
 		return [
 			(leftSideX, 0),
 			(rightSideX, 0),
 			(0, topSideY),
 			(0, bottomSideY),
 		]
 
-	def getCoordinatesAndControls(self, glyphName, ttFont, defaultVerticalOrigin=None):
+	def getCoordinatesAndControls(self, glyphName, hMetrics, vMetrics=None):
```

Comment:
I have external code that uses this method. While it's easy enough to adapt (and I'm willing to), this seems to be a public method, and this change is not b/w compatible.

Reply:
would it be possible to add a new method and keep this one but with a deprecation warning? I know it's difficult to come up with a good alternative name. Maybe just `getCoordinates`?

Reply:
Okay, I'll put the old one back. If doing that I might as well leave the old `getPhantomPoints` alone...

For names.. I don't like just `getCoordinates` given the phantomPoint dance this does. I'm happy to add `_getPhantomPoints` and `_getCoordinatesAndControls` to mark them internal so we can adapt as needed.

Reply:
> I have external code that uses this method.

Thanks @justvanrossum. I'm curious how you use it. Do you use the control data as well? The phantom points?

Reply:
I suppose I have to leave alone the old `setCoordinates` as well then.

Reply:
@behdad: my code is pretty much a copy of this fragment in mutator.py:

https://github.com/fonttools/fonttools/blob/3cfc87be711756baede6059d088cc2a2ebd20605/Lib/fontTools/varLib/mutator.py#L196-L207

I need to instantiate an individual composite glyph. I don't use the controls, and strip the phantom points.

I can't use instancer's `instantiateGvarGlyph()` as this modifies the `glyf` table: all I need at this point are the instantiated component offsets.

A public function doing something like "give me the instantiated coordinates as a GlyphCoordinates array" would be helpful for my case.

Reply:
I also have a copy of `instantiateGvarGlyph()` in a tool to instantiate a glyph at a certain point without changing the font :O

Reply:
Understood. Let's add a designed API for instantiation; ideally as methods of an instancer object.

Reply:
Cosimo said instancer is made for font mutation. Changing that would probably be an involved refactoring.

Reply:
As it happens I've been looking into that code. I like to see L4 subsetting happening. To me, it's a waste if we don't use the same code to back GUI tools like FontGoggle and others.

Can you open an issue please, expanding on your use-case, and hopefully we get Just's as well. Then we can design a new module to lay under or replace `mutator` and `instancer`.